drivers/lg/displays/ls5.cr

```diff
@@ -0,0 +1,297 @@
+    queue.delay = 150.milliseconds
+    transport.tokenizer = Tokenizer.new(Bytes[DELIMITER])
+    on_update
+  end
+
+  def on_update
+    @rs232 = setting(Bool, :rs232_control)
+    @id_num = setting(Int32, :display_id)
+    @id = @id_num.to_s.rjust(2, '0')
+  end
+
+  def connected
+    wake_on_lan
+    no_signal_off
+    auto_off
+    local_button_lock
+    pm_mode
+    schedule.every(50.seconds, true) do
+      do_poll
+    end
+  end
+
+  def disconnected
+    schedule.clear
+  end
+
+  enum Command
+    Power           = 0x61 # 'a'
+    Input           = 0x62 # 'b'
+    AspectRatio     = 0x63 # 'c'
+    ScreenMute      = 0x64 # 'd'
+    VolumeMute      = 0x65 # 'e'
+    Volume          = 0x66 # 'f'
+    Contrast        = 0x67 # 'g'
+    Brightness      = 0x68 # 'h'
+    Sharpness       = 0x6B # 'k'
+    AutoOff         = 0x6E # 'n'
+    LocalButtonLock = 0x6F # 'o'
+    WakeOnLan       = 0x77 # 'w'
+    NoSignalOff     = 0x67 # 'g'
+    PmMode          = 0x6E # 'n'
+  end
+  {% for name in Command.constants %}
+    @[Security(Level::Administrator)]
+    def {{name.id.underscore}}?(priority : Int32 = 0)
+      do_send(Command::{{name.id}}, 0xFF, priority: priority, name: {{name.id.underscore.stringify}} + "_status")
+    end
+  {% end %}
+
+  def power(state : Bool, broadcast : String? = nil)
+    if state
+      if @rs232
+        do_send(Command::Power, 1, name: "power", priority: 99)
+      else
+        wake(broadcast || @last_broadcast)
+      end
+    end
+    # To power on, unmute the display
+    # To power off, mute the display
+    mute(!state) if self[:connected]
+  end
+
+  def hard_off
+    do_send(Command::Power, 0, name: "power", priority: 99)
+  end
+
+  def switch_to(input : Input, **options)
+    do_send(Command::Input, input.value, 'x', name: "input", delay: 2.seconds)
+  end
+
+  def mute(
+    state : Bool = true,
+    index : Int32 | String = 0,
+    layer : MuteLayer = MuteLayer::AudioVideo
+  )
+  mute_video(state) if layer.video? || layer.audio_video?
+  mute_audio(state) if layer.audio? || layer.audio_video?
+  end
+
+  def mute_video(state : Bool = true)
+    state = state ? 1 : 0
+    do_send(Command::ScreenMute, state, name: "mute_video")
+  end
+
+  def mute_audio(state : Bool = true)
+    # Do nothing if already in desired state
+    return if (self[:audio_mute]?.try &.as_bool) == state
+    state = state ? 0 : 1
+    do_send(Command::VolumeMute, state, name: "mute_audio")
+  end
+
+  enum Ratio
+    Square  = 0x01
+    Wide    = 0x02
+    Zoom    = 0x04
+    Scan    = 0x09
+    Program = 0x06
+  end
+  def aspect_ratio(ratio : Ratio)
+    do_send(Command::AspectRatio, ratio.value, name: "aspect_ratio", delay: 1.second)
+  end
+
+  def do_poll
+    if @rs232
+      power?
+      if self[:hard_power]?.try &.as_bool
+        screen_mute?
+        input?
+        volume_mute?
+        volume?
+      end
+    elsif self[:connected]?.try &.as_bool
+      screen_mute?
+
+      if @id_num == 1
```

Comment:
No idea why actually as I've copied this over from the ruby driver. I can get rid of this but it might have been there for a reason?

Not exactly of the difference between communicating with a device that's hooked up via rs232 compared to one that's not regarding this bit of code as well which I also carried over from the ruby driver
https://github.com/PlaceOS/drivers/blob/19d0468789bbed049f8f42aa2d84de565404a631/drivers/lg/displays/ls5.cr#L145-L160

Reply:
I think from a Crystal engine point of view, `self[:connected] == JSON::Any.new(true / false)`
So `JSON::Any.new(false)` evaluates to `true` so need to use `.raw` or move it into the expected type via `as_bool` (will raise an error if an Int etc)

The real answer is to track `@connected` state via an instance variable and just update the public state as required (much more performant too)

Reply:
Thanks @stakach , I've made these commits.

https://github.com/PlaceOS/drivers/pull/61/commits/d5c37cbd170c895a8be5cbff624ee28661b3758f

https://github.com/PlaceOS/drivers/pull/61/commits/2d7d4fc2158a4f93510527dac15f827f52a6b1c7